### PR TITLE
Type a ranged value for filter_var(FILTER_VALIDATE_INT)

### PR DESCRIPTION
**Pull request by zonuexe**
closes https://github.com/phpstan/phpstan/issues/7494

### CODE REVIEW

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -91,6 +97,24 @@ private function getFilterTypeMap(): array
...
+	private function getFilterTypeOptions(): array
+	{
+		if ($this->filterTypeOptions !== null) {
+			return $this->filterTypeOptions;
+		}
+
+		$this->filterTypeOptions = [
+			$this->getConstant('FILTER_VALIDATE_INT') => ['min_range', 'max_range'],
+			// PHPStan does not yet support FloatRangeType
+			// $this->getConstant('FILTER_VALIDATE_FLOAT') => ['min_range', 'max_range'],
```
Comment (zonuexe) [inline src/Type/Php/FilterVarDynamicReturnTypeExtension.php:111]: I don't plan to use it immediately, but it is possible to make `FloatRangeType` with `-INF <= float <= INF`.

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php [hunk 2]
```diff
@@ -129,23 +153,39 @@ public function getTypeFromFunctionCall(
...
-			if ($inputType->isNonEmptyString()->yes()
-				&& $type instanceof StringType
-				&& !$this->canStringBeSanitized($filterValue, $flagsArg, $scope)) {
-				$type = new IntersectionType([$type, new AccessoryNonEmptyStringType()]);
-			}
+		$typeOptionNames = $this->getFilterTypeOptions()[$filterValue] ?? [];
+		$otherTypes = $this->getOtherTypes($flagsArg, $scope, $typeOptionNames);
+
+		if ($inputType->isNonEmptyString()->yes()
+			&& $type instanceof StringType
```
Comment (ondrejmirtes) [inline src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: `$type->isString()->yes()`

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php [hunk 3]
```diff
@@ -129,23 +153,39 @@ public function getTypeFromFunctionCall(
...
+			&& !$this->canStringBeSanitized($filterValue, $flagsArg, $scope)) {
+			$type = new IntersectionType([$type, new AccessoryNonEmptyStringType()]);
```
Comment (ondrejmirtes) [inline src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: `TypeCombinator::intersect()` instead please

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php [hunk 4]
```diff
@@ -129,23 +153,39 @@ public function getTypeFromFunctionCall(
...
+				$type = $otherTypes['range'];
 			}
 		}
 
+		if ($exactType !== null) {
+			unset($otherTypes['default']);
+		}
+
+		if (isset($otherTypes['default']) && $otherTypes['default']->isSuperTypeOf($type)->no()) {
+			$type = new UnionType([$type, $otherTypes['default']]);
```
Comment (ondrejmirtes) [inline src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: `TypeCombinator::union()` please

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php [hunk 5]
```diff
@@ -168,43 +208,74 @@ private function determineExactType(Type $in, int $filterValue): ?Type
...
+				$range['max'] = $typeOptions['max_range']->getValue();
+			} elseif ($typeOptions['max_range'] instanceof IntegerRangeType) {
+				$range['max'] = $typeOptions['max_range']->getMax();
+			}
+		}
+
+		if (isset($range['min']) || isset($range['max'])) {
+			$min = isset($range['min']) && is_int($range['min']) ? $range['min'] : null;
+			$max = isset($range['max']) && is_int($range['max']) ? $range['max'] : null;
+			$otherTypes['range'] = new IntegerRangeType($min, $max);
```
Comment (ondrejmirtes) [inline src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: `IntegerRangeType::fromInterval()` please

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php (hunk 3 above)
Comment (zonuexe) [reply src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: Fixed https://github.com/phpstan/phpstan-src/pull/1443/commits/fec2ef66a70218522bf5954338b3aa84dd8b3a20

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php (hunk 5 above)
Comment (zonuexe) [reply src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: Fixed https://github.com/phpstan/phpstan-src/pull/1443/commits/09ae19764dab721c0ff26ef34ee363d089928316

File: src/Type/Php/FilterVarDynamicReturnTypeExtension.php (hunk 2 above)
Comment (zonuexe) [reply src/Type/Php/FilterVarDynamicReturnTypeExtension.php]: Fixed https://github.com/phpstan/phpstan-src/pull/1443/commits/e301f6dbaea2eeed2a740d862c37602b940387d8

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**